### ISSUE TITLE
DEPRECATION WARNING: Class level methods will no longer inherit scoping from `create!`

### ISSUE THREAD

**Issue by equivalent** [rails/rails#38881]
I do understand that issue I'll describe bellow is due to this PR change https://github.com/rails/rails/pull/32380  in Rails 6 and that it has good reason. What I want to do in this issue is to extend the discussion on valid workarounds for future.

### description of situation

Given:

```ruby
class Blog < ActiveRecord::Base
  has_many :links
end

class Link < ActiveRecord::Base
   scope  :youtube, -> { where source: 'youtube' }
   scope  :netflix, -> { where source: 'netflix' }
   scope  :basecamp, -> { where source: 'basecamp' }
end

class SomeService
   def self.scenario1
       blog = Blog.last
       blog.links.youtube.create!
   end
end
```

When I do

```ruby
SomeService.scenario1
```

Then I Get Deprecation warning:

```
DEPRECATION WARNING: Class level methods will no longer inherit scoping from `create!` in Rails 6.1. To continue using the scoped relation, pass it into the block directly. To instead access the full set of models, as Rails 6.1 will, use `Link.unscoped`, or `Link.default_scoped` if a model has default scopes. (called from some_method_name at ..... )
```

### the issue

The easiest workaround I came up with  that will result not displaying the deprecation message is:

```ruby
class SomeService
   def self.scenario2
       blog = Blog.last
       link = blog.links.youtube.build
       link.save!
       link
   end
end

SomeService.scenario2
```

But then what is the point of not allowing `create!` to do the scoping when `blog.links.youtube.build` is able to do it. 

My question would be: is this valid workaround for future ? (E.g Will in rails 6.x this  get deprecated as well ? ) If no what is more valid solution for this situation ?

### System configuration
**Rails version**: 6.0.2.2

**Ruby version**: 2.6.5

**Comment by rafaelfranca** [comment]
@kamipo I think we should not show deprecation in this case. The reason why we deprecated was to avoid callbacks to inherit the scope, but there is no callback in this use case.

@equivalent
Could you create a reproduction script with this template https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb?

**Comment by kamipo** [comment]
I can't reproduce this issue, and we have similar test case that association scope chain with create/create!:

https://github.com/rails/rails/blob/72d293601346f708082d9ecfcd9b99b987f64783/activerecord/test/cases/associations/has_many_associations_test.rb#L252-L256

I suppose your app have any (explicit or implicit) callbacks on the model, potentially be affected by scoping leak, similar with #38241.

**Comment by rafaelfranca** [comment]
That is what I thought, this is why I asked the reproduction script/

**Comment by equivalent** [comment]
@rafaelfranca  that's interesting I've created the replication script and no deprecation msg 

https://gist.github.com/equivalent/52e3713ec9c35aed26a463ccc28d0617

run on Rails 6.0.2.2 and 6.0.2.1 same result -> no Deprication mgs :thinking:

**Comment by equivalent** [comment]
so it's strange I cannot reproduce it with the script and don't see it as such a big deal to create dummy application to reproduce it  and honestly it's not a big deal as I can use `.build.tap{ |r| r.save! }` and it will do same thing  (**so ok to close this ticket**)

I really just wanted to know if this is still valid approach (it is )  the deprecated message is just detail


that being said this really does happen 

before:

![Screenshot from 2020-04-07 15-12-19](https://user-images.githubusercontent.com/721990/78673550-95d61a00-78e2-11ea-9326-8f3357368b1d.png)
![Screenshot from 2020-04-07 15-11-37](https://user-images.githubusercontent.com/721990/78673559-98d10a80-78e2-11ea-81de-e3fcbf0038fc.png)

after 

![Screenshot from 2020-04-07 15-16-30](https://user-images.githubusercontent.com/721990/78673685-c7e77c00-78e2-11ea-9798-0c61dbe27d83.png)
![Screenshot from 2020-04-07 15-16-48](https://user-images.githubusercontent.com/721990/78673722-d33aa780-78e2-11ea-95bc-48cd76fc0a03.png)

**Comment by kaspth** [comment]
What @kamipo has been saying is that there’s possibly some conditions in the scopes that you use that force the deprecation warning. Can you show those?

**Comment by equivalent** [comment]
@kaspth yep sure

scopes are defined like this:

```ruby
# app/model/magic_link.rb
class MagicLink < ApplicationRecord

  scope :presentation_shared,
    -> { where(type: MagicLinkTypes::PresentationShared.name) }
   scope :confirm_email_identities,
    -> { where(type: MagicLinkTypes::ConfirmEmailIdentity.name) }
  scope :confirm_email_changes,
    -> { where(type: MagicLinkTypes::ConfirmEmailChangeEmailIdentity.name) }
  scope :reset_password_email_identities,
    -> { where(type: MagicLinkTypes::ResetPasswordEmailIdentity.name) }

  # ...
end
```

```ruby
# app/model/shared_presentation.rb
class SharedPresentation < ApplicationRecord
  has_many :magic_links, as: :resource, dependent: :destroy
  # ...
end
```

```ruby
# app/model/magic_link_types/shared_presentation.rb
module MagicLinkTypes
  class PresentationShared < Base
      # this is just a business logic class.
      #  if you wondering if the class method `self.name` has any override: nope,  Standard Class.name behavior from Ruby
  end
end
```

**Comment by rafaelfranca** [comment]
This deprecation usually happens in callbacks. Do you have any `MagicLink.something` call in a callback?

**Comment by equivalent** [comment]
only callbacks are:

```ruby
class MagicLink < ApplicationRecord
  before_create :generate_and_set_token
  before_create :set_expire_at_hook

 def generate_and_set_token
    self.token = SecureRandom.hex(50) unless token
  end

  def set_expire_at_hook
    self.expire_at = token_type.class.expire_at unless expire_at
  end

  def token_type
    @token_type ||= type.constantize.new(self)
  end
end

module MagicLinkTypes
  class PresentationShared < Base
    def self.expire_at
      2.days.from_now
    end
  end
end

```

**Comment by rafaelfranca** [comment]
I think you will have to inspect the deprecation backtrace to find where the scope access in the class is being called.

**Comment by equivalent** [comment]
we don't have to dig deeper into my application.

What I've really wanted to achieve with this ticket is to ensure that by using  scopes to build/create my objects I will not introduce something I will have to refactor my code in Rails 6.2 

so it's fine we can close this ticket if you guys want to  (In my opinion we can leave this open as others may encounter this too)

**Comment by kamipo** [comment]
You can find a callback by `puts caller` as I do often:

```diff
diff --git a/activerecord/lib/active_record/scoping/named.rb b/activerecord/lib/active_record/scoping/named.rb
index 2bed5226b7..5282311aaa 100644
--- a/activerecord/lib/active_record/scoping/named.rb
+++ b/activerecord/lib/active_record/scoping/named.rb
@@ -32,6 +32,7 @@ def all
                 in Rails 6.1. To continue using the scoped relation, pass it into the block directly.
                 To instead access the full set of models, as Rails 6.1 will, use `#{name}.default_scoped`.
               MSG
+              puts caller
             end
 
             if self == scope.klass
```

Anyway, I'll close this since I think this is not an issue for Rails code base.

**Comment by estani** [comment]
Hmm... I'm having the same issue and this is the first thread that describes the problem I'm seeing.

As a take away I understand:
- `has_many` + `find_or_create_by` should indeed work together (because is not clear to me, as it defines a scope that is neither `default_scoped` or `unscoped`, this will still be supported in Rails 6.1)
- The problem is/will be related to callbacks.

tl;dr

Should we avoid using `find_or_create_by` on `has_many` relations in callbacks?


What we do is, after a model is saved, a callback might trigger the creation of a different model, via a has_many relation. There are a few tests going through this spot, and not everyone throws the deprecation warning (even more strange). Only the callbacks from `after_create` seem to cause this (from `after_update` or `after_destroy` they don't, and all call the same method) but that might be a red herring...

I also found a workaround for the warning similar to the one mentioned in the first post, which as was already said makes little sense. If the scope of the `has_many` relation is lost, I will have a problem when updating to 6.1

Some code to depict the work_around. (sometimes) Throwing the deprecation warning:
```
parent.children.find_or_create_by(attr: value)
```

Never throwing the deprecation warning:
```
parent.children.scoping { parent.children.find_or_create_by(attr: value) }
```

What I will do now, after reading this thread (I think is more resilient):
```
Child.default_scoped.find_or_create_by(parent:parent , attr: value)
```

By the way, I recommend using `ActiveSupport::Deprecation.debug = true` to debug deprecation warnings.
